lib/posix-process: Define `libc` wrappers for `clone`/`vfork`

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

NOTE: This PR begins with several minor fixes/improvements as prerequisites

`clone`:
In order to successfully call the `clone` system call through a libc
wrapper we need to be able to have an assembly sequence that both
translates the arguments passed by the libc wrapper to those of the
underlying syscall, since their signatures differ, as well as redirect
the parent and child accordingly upon exit from said syscall: child
must run its requested function and argument and parent must return
to the clone caller with an unscathed register context.

Said assembly sequence must be aware of the signature differences
between architectures, e.g. ARM64 vs x86_64 signatures of the
clone system call.

`vfork`:
As other libc's also state, it is impossible to write the libc
wrapper of the `vfork` system call in C since the child might
end up reusing and overwriting the parent's return address when it
pops it and then calls `execve`, leaving the parent to return to an
invalid state - typically would overwrite it with the return address
from execve, making it look as if the parent returned from execve.

Other libc's solve this by writing the libc wrapper as something
among the lines of:
popq    %<register that the syscall guarantees to preserve>
movq    $SYS_vfork, %rax
syscall
pushq   %<register we popped return address into 3 lines above>
ret

On ARM64 as well as others, clone() with the flags
`SIGCHLD | CLONE_VM | CLONE_VFORK` is called instead but let's just
still use our `vfork` implementation since it essentially also does
the same exact thing.

However, as a Unikernel with function calls for system calls we do
not need to care about any of that, as our execenv prologue can store
and restore everything without touching the stack. Thus, we can write
our native libc wrappers as a basic jump/branch to said prologue.

mogasergiu committed:
lib/syscall_shim/arch/x86_64: Save proper rsp in execenv prologue

In the execenv prologue meant for native builds we try to mimic the
context save/restore that would happen following a syscall instruction
but in the case of a direct call instruction. This means that the rsp
on entry is actually 8 bytes less than the rsp we are supposed to show
to actual users of this execenv. To cope with this, after pushing the
original rsp do an addition of 8 so that children (e.g. vfork) that may
inherit this context have the proper rsp. Lastly, because of this, upon
exiting the execenv assembly wrapper we must ensure that the context
whose execenv we store/restore is using the proper rsp as well by
undoing aforementioned addition, since it actually returns like a normal
function through ret.

This bug hasn't been caught before because we've only been using this
in the context of the clone syscall for native builds. Unlike vfork,
in the case of clone, the children typically begin execution with a
brand new stack instead of reusing and mimicking that of the parent.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim: Add non-LL execenv syscall registration variant

Based on UK_LLSYSCALL_R_E_DEFINE and provides a libc-style wrapper
in case UK_LIBC_SYSCALLS is enabled.

(Similar to its UK_SYSCALL_R_DEFINE, non-execenv variants)

Checkpatch-Ignore: MACRO_ARG_REUSE
Checkpatch-Ignore: COMPLEX_MACRO
Checkpatch-Ignore: SPACING
Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim: Remove unnecessary `__used` from execenv variants

The `__used` attribute in not necessary as we are fine with any of
these potentially being optimised away.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/posix-process: Register `execve` using non-LL variant

Use the non-LL variant of execenv-based syscall registration for
execenv so that we will automatically have a corresponding libc
wrapper defined.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim: Add `uk_syscall_do_` syms for execenv registrations

Because of the unnecessary conditional during syscall symbols
generation, execenv syscall registrations would not benefit from
having their `uk_syscall_do_` symbol generated automatically.
Fix this by removing this conditional.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
arch: Define regular function calling convention register arguments

Define the macros that one could use to access the registers within
`struct __regs` that represent the ABI/calling convention specific
that contain the arguments (and the return value).

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/arm64: Store caller's link register

So far for storing LR and ELR_EL1 we have been using the current
link register for both instead. This does not reflect reality when
it comes to LR as we should instead store the LR that the caller
would know. To achieve this, do a brief level 1 stack unwinding and
fetch caller's LR from the top of its frame.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/x86_64: Restore `RIP` from the auxiliary stack

Before this patch, we would simply rely on the original pushed RIP
following the call instruction that got to our assembly wrapper.
However this may not be the same in cases such as those of the clone
or vfork system calls if the child were to reuse the stack: the child
could pop the return address before the parent gets the chance to do it
and even call some other functions (like execve), overwriting whatever
previously was at the bottom of the stack that the parent had prior
to invoking the system call.

To solve this, simply use the RIP pushed at the beginning of the wrapper
instead of assuming the bottom of the stack is untouched.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/posix-process: Add `libc` system call wrapper for `clone`

In order to successfully call the clone system call through a libc
wrapper we need to be able to have an assembly sequence that both
translates the arguments passed by the libc wrapper to those of the
underlying syscall, since their signatures differ, as well as redirect
the parent and child accordingly upon exit from said syscall: child
must run its requested function and argument and parent must return
to the clone caller with an unscathed register context.

Said assembly sequence must be aware of the signature differences
between architectures, e.g. ARM64 vs x86_64 signatures of the
clone system call.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/posix-process: Fix `vfork` signature

A minor oversight during merging of the `vfork` functionality ended
up defining `vfork` with an invalid signature that takes two arguments.

Fix this by defining `vfork` as a syscall that takes no arguments.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/posix-process: Define `libc` wrappers for `vfork`

It is impossible to write the libc wrapper of the `vfork` system call
in C since the child might end up reusing and overwriting the parent's
return address when it pops it and then calls `execve`, leaving the
parent to return to an invalid state - typically would overwrite it
with the return address from execve, making it look as if the parent
returned from execve.

Userspace libc's solve this by writing the libc wrapper as something
among the lines of:
popq	%<register that the syscall guarantees to preserve>
movq	$SYS_vfork, %rax
syscall
pushq	%<register we popped return address into 3 lines above>
ret

On ARM64 as well as others, clone() with the flags
`SIGCHLD | CLONE_VM | CLONE_VFORK` is called instead but let's just
still use our `vfork` implementation since it essentially also does
the same exact thing.

However, as a Unikernel with function calls for system calls we do
not need to care about any of that, as our execenv prologue can store
and restore everything without touching the stack. Thus, we can write
our native libc wrappers as a basic jump/branch to said prologue.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu commented:
Depends-on: [#86](https://github.com/unikraft/lib-musl/pull/86)

andreittr commented on lib/syscall_shim/arch/x86_64/include/arch/syscall_prologue.h:
```diff
@@ -47,8 +47,16 @@
+		" * pushes the address the ret instruction is\n\t"	\
+		" * supposed to return to. This means that to truly\n\t"\
+		" * mimic a trap/syscall we must store/restore\n\t"	\
+		" * the rsp we were given, plus 8.\n\t"			\
```
The code makes sense, but I had to read the comments through a couple of times before getting what's going on. It might parse easier if this large comment were moved earlier, where we save the current sp (`movq %rsp, %r11`) and reworded to clearly explain what's going on with the stack. Something like

"Our stack top now contains a return address pushed by `call`; this must be ignored when saving the stack pointer to the interrupt return structure, but taken into account when we actually return execution", with smaller, single-single sentence reminders here when we `addq`, and later when we `subq`.

(If my rewording completely missed the point, then I believe it's evidence enough that the comment could do better :smile: )

andreittr commented on lib/syscall_shim/arch/x86_64/include/arch/syscall_prologue.h:
```diff
@@ -111,6 +119,11 @@
+		"/* Now subtract those 8 bytes we previously added\n\t"	\
+		" * because the caller/ret do not expect the rsp to\n\t"\
+		" * not point to the actual return address.\n\t"	\
+		" */\n\t"						\
```
Tangential to above, could use simplification & removal of double-negative: "[ ... ] do not expect [...] to not point":

"Adjust saved stack to original value; `ret` expects return address on top of stack"

andreittr commented on arch/x86/x86_64/include/uk/asm/lcpu.h:
```diff
@@ -66,6 +66,22 @@
+#define __fn_rarg4		r8
+#define __fn_rarg5		r9
+
+#define __fn_rret		rax
```
Can I assume we are explicitly discounting the case of return value in `%rax:%rdx`?
AFAICT these macros are used for syscalls, and all syscalls return at most a machine word, so _should be fine_.

mogasergiu replied:
Doesn't hurt to add it. Will do.

andreittr commented on lib/posix-process/clone.c:
```diff
@@ -593,12 +593,94 @@ UK_LLSYSCALL_R_E_DEFINE(int, clone,
+	"/* Mark the outermost frame (avoid unwinding beyond this)\n\t"
+	"xorq	%rbp, %rbp\n\t"
+	"call	*%rbx\n\t"
+	"/* If managed to return from fn(arg) we must call exit(x0) */\n\t"
```
What is `x0` in `exit(x0)`?
- It is a typo and intended is `exit(0)`? In that case you need a `xor %edi, %edi` before the call
- Is it intended as `exit(rax)` (i.e., return code of `fn`)? Then you need a `mov %rax, %rdi` before the call.

mogasergiu replied:
Yup, the latter, good catch.

andreittr commented on lib/posix-process/clone.c:
```diff
@@ -593,12 +593,94 @@ UK_LLSYSCALL_R_E_DEFINE(int, clone,
+	"/* The child should now just call the saved fn(arg) */\n\t"
+	"movq	%rbp, %rdi\n\t"
+	"/* Mark the outermost frame (avoid unwinding beyond this)\n\t"
+	"xorq	%rbp, %rbp\n\t"
```
You can save a byte by doing `xor %ebp, %ebp`, and in general making use of the defined behavior of zeroing out the upper half of 64bit regs when their lower halves are written to.
```
   0:	48 31 ed             	xor    %rbp,%rbp
   3:	31 ed                	xor    %ebp,%ebp
```

mogasergiu replied:
I've been told this so many times and I still lose track of it when I am writing assembly :rofl: .
Will do, thanks! :+1:

andreittr reviewed (commented):
On 1st pass looks ok modulo some comments.

mogasergiu commented:
Rebase force push

michpappas commented on lib/posix-process/Config.uk:
```diff
@@ -26,18 +26,19 @@ config LIBPOSIX_PROCESS_SIGNAL
 	select LIBUKLOCK
 
-config LIBPOSIX_PROCESS_CLONE
+menuconfig LIBPOSIX_PROCESS_CLONE
```
This is a bit out of scope of this PR and I'm about to open a PR that reworks the config options of `libposix-process`, which - among others - deprecates `LIBPOSIX_PROCESS_CLONE` in favor of `LIBPOSIX_PROCESS_MULTITHREADING`, and also moves `PREFER_CHILD` under `LIBPOSIX_PROCESS_MULTITHREADING` too. So maybe we can save that change for that PR?

mogasergiu replied:
OK I can drop this commit.

michpappas commented on lib/posix-process/clone.c:
```diff
@@ -593,12 +593,95 @@ UK_LLSYSCALL_R_E_DEFINE(int, clone,
-}
+	  ... /* pid_t *parent_tid, void *tls, pid_t *child_tid */);
+
+#if CONFIG_ARCH_X86_64
```
Is there any reason these are not under `lib/posix-process/arch/<arch>/clone.S`?

mogasergiu replied:
Just laziness. In fact, Andrei also suggested I do that. Makes sense to me. It would also make source debugging easier.

michpappas commented on lib/posix-process/vfork.c:
```diff
@@ -16,6 +16,25 @@
 
 #include "process.h"
 
+#if UK_LIBC_SYSCALLS
```
Same as in `clone()`, these should be under `lib/posix-process/arch/<arch>/vfork.S`.

michpappas commented on lib/posix-process/clone.c:
```diff
@@ -593,12 +593,95 @@ UK_LLSYSCALL_R_E_DEFINE(int, clone,
+	"/* Save fn and arg in callee-saved registers, but first make sure\n\t"
+	" * we can also restore them before returning in the parent.\n\t"
+	" */\n\t"
+	"mov	x20, x0\n\t"
```
First save and then corrupt?

mogasergiu replied:
Yes I like my crash dumps confused.

... No idea how I ended up misplacing that `stp`. I made the same mistake in the musl PR. Will change there as well. :+1:

michpappas reviewed (changes requested):
Nice set of changes! In addition to the inline comments, I would suggest that you added a noticable banner with cross-referencing comments in the duplicated code here and in musl to make sure that anyone updating one keeps the other one in sync too.

On the asm wrappers I focused on arm64 that I'm more familiar with, as @andreittr covered x86_64 already.

Some nits below:

As general remark, as I have also noticed in other PRs, please refrain from using markdown in the title of PRs, as that causes the PR formatting check to fail as you can see below. To make things even worse, the script ignores edits, so even if it's corrected, the error will persist 🤦🏼‍♂️ 

Finally some small things I noticed in the commit messages:

In `lib/syscall_shim/arch/x86_64: Save proper rsp in execenv prologue`:

> to actual users of thie execenv
"to actual users of this execenv"

In `lib/posix-process: Define libc wrappers for vfork`:

> As other libcs also state

The code this patch applies to is not a libc though. It would be nice to provide links to the relevant code as references. A possible rephrase could be:

"As stated in the implementation of libc [1] and musl [2]..."

mogasergiu committed:
lib/syscall_shim: Make `UK_LIBC_SYSCALLS` visible in asm files

Define `UK_LIBC_SYSCALLS` outside the `!__ASSEMBLY__` guard so that
assembly files can see it.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

michpappas reviewed (approved):
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

andreittr reviewed (approved):
I much rather like separate asm files for the wrappers, nice change.

Shame `UK_LIBC_SYSCALLS` is not a true config option but computed by `syscall_shim` :disappointed: ; we could've saved ourselves an `#ifdef` and put all conditions for the asm files entirely in the Makefile. Oh well :smile: 

Reviewed-by: Andrei Tatar <andrei@unikraft.io>

michpappas reviewed (approved):
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

andreittr commented on lib/posix-process/arch/x86_64/clone.S:
```diff
@@ -0,0 +1,74 @@
+	/* child_tid: this is passed on the stack as it does not fit
+	 * in the registers used for the arguments
+	 */
+	movq	24(%rsp), %r10
```
Oof, having to place the 4th argument of a function call in the register used by syscalls will 100% be something that trips future readers up, some of whom might decide to "fix" it to `rcx`.
I (vaguely) understand it's to do with execenv syscalls being special and requiring binary args, but I believe future people deserve a very explicit warning saying that `r10` is here on purpose and why.

andreittr reviewed (commented):
Code looks fine, shame we didn't catch the wrong stack address in previous review :disappointed: .
I've got only one documentation/comment request.

andreittr reviewed (approved):
Thanks for the changes, looks good!

Reviewed-by: Andrei Tatar <andrei@unikraft.io>

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
